algorithms/kernel/dtrees/forest/regression/oneapi/df_regression_train_dense_default_oneapi_impl.i

```diff
@@ -0,0 +1,1245 @@
+services::Status RegressionTrainBatchKernelOneAPI<algorithmFPType, method>::computeBestSplitByHistogram(
+    const UniversalBuffer & nodeHistogramList, UniversalBuffer & selectedFeatures, size_t nSelectedFeatures, UniversalBuffer & nodeList,
+    UniversalBuffer & nodeIndices, size_t nodeIndicesOffset, UniversalBuffer & binOffsets, UniversalBuffer & impList,
+    UniversalBuffer & nodeImpDecreaseList, bool updateImpDecreaseRequired, size_t nNodes, size_t nMaxBinsAmongFtrs, size_t minObservationsInLeafNode,
+    algorithmFPType impurityThreshold)
+{
+    DAAL_ITTNOTIFY_SCOPED_TASK(compute.computeBestSpitByHistogramLevel);
+
+    services::Status status;
+
+    auto & context = services::Environment::getInstance()->getDefaultExecutionContext();
+
+    auto & kernel = kernelComputeBestSplitByHistogram;
+
+    {
+        KernelArguments args(13);
+        args.set(0, nodeHistogramList, AccessModeIds::read);
+        args.set(1, selectedFeatures, AccessModeIds::read);
+        args.set(2, (int)nSelectedFeatures);
+        args.set(3, binOffsets, AccessModeIds::read);
+        args.set(4, nodeList, AccessModeIds::readwrite); // nodeList will be updated with split attributes
+        args.set(5, nodeIndices, AccessModeIds::read);
+        args.set(6, nodeIndicesOffset);
+        args.set(7, impList, AccessModeIds::write);
+        args.set(8, nodeImpDecreaseList, AccessModeIds::write);
+        args.set(9, (int)updateImpDecreaseRequired);
+        args.set(10, (int)nMaxBinsAmongFtrs);
+        args.set(11, (int)minObservationsInLeafNode);
+        args.set(12, impurityThreshold);
+
+        const size_t numOfSubGroupsPerNode = 4; //add logic for adjusting it in accordance with nNodes
+        size_t localSize                   = _preferableSubGroup * numOfSubGroupsPerNode;
+
+        KernelRange local_range(localSize, 1);
+        KernelRange global_range(localSize, nNodes);
+
+        KernelNDRange range(2);
+        range.local(local_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+        range.global(global_range, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+
+        context.run(range, kernel, args, &status);
+        DAAL_CHECK_STATUS_VAR(status);
+    }
+
+    return status;
+}
+template <typename algorithmFPType, decision_forest::regression::training::Method method>
+services::Status RegressionTrainBatchKernelOneAPI<algorithmFPType, method>::computeBestSplit(
+    const UniversalBuffer & data, UniversalBuffer & treeOrder, UniversalBuffer & selectedFeatures, size_t nSelectedFeatures,
+    UniversalBuffer & response, UniversalBuffer & nodeList, UniversalBuffer & binOffsets, UniversalBuffer & impList,
+    UniversalBuffer & nodeImpDecreaseList, bool updateImpDecreaseRequired, size_t nFeatures, size_t nNodes, size_t minObservationsInLeafNode,
+    algorithmFPType impurityThreshold)
+{
+    services::Status status;
+    const size_t nHistProps     = 3;
+    const size_t nNodesGroups   = 3;
```

Comment:
magic 🧙

Reply:
It is better to move such magic constants into class declaration in header file.

Reply:
yes, will do

Reply:
moved them in header